eval.py

```diff
@@ -0,0 +1,217 @@
+from scipy.interpolate import interp1d
+from matplotlib import rcParams
+import matplotlib.pyplot as plt
+from matplotlib.ticker import PercentFormatter
+rcParams["figure.figsize"] = [9.0, 6.0]
+
+
+class Evaluator(object):
+
+    """Incremental Test Outcome Object."""
+
+    def __init__(self, pred_out, name, exclude_other=False):
+        """Initialization.
+
+        :pred_out: prediction outcome from the model
+        :name: experiment name
+        :exclude_other: whether to exclude `O` tags when calculating slot f1 scores
+        """
+
+        self.logger = ColoredLog(__name__)
+        self.name = name
+        self.exclude_other = exclude_other
+        self.test_set = {}
+        self.error = defaultdict(list)
+        self.count = defaultdict(int)
+        self.length_map = defaultdict(list)
+        for i in range(len(pred_out["sorted_ids"])):
+            sent_id = pred_out["sorted_ids"][i][:10]
+            sub_id = pred_out["sorted_ids"][i][11:]
+            if sent_id not in self.test_set:
+                self.test_set[sent_id] = defaultdict(dict)
+            if sub_id != "full":
+                sub_id = int(sub_id)
+                self.test_set[sent_id][sub_id]["gold_int"] = pred_out["golden"][i]
+                self.test_set[sent_id][sub_id]["pred_int"] = pred_out["pred"][i]
+                if pred_out["golden"][i] != pred_out["pred"][i]:
+                    self.error["intent"].append((sent_id, sub_id))
+                self.test_set[sent_id][sub_id]["gold_slot"] = pred_out["golden_slot"][i]
+                self.test_set[sent_id][sub_id]["pred_slot"] = pred_out["pred_slot"][i]
+                if not (np.array(pred_out["golden_slot"][i]) == np.array(pred_out["pred_slot"][i])).all():
+                    self.error["slot"].append((sent_id, sub_id))
+                self.test_set[sent_id][sub_id]["text"] = pred_out["text"][i]
+            else:
+                length = len(pred_out["golden_slot"][i]) - 2
+                self.count[length] += 1
+                self.length_map[length].append(sent_id)
+
+        self.intent_set = self.generate_intent_map(pred_out["golden"])
+        self.slot_set = self.generate_slot_map(pred_out["golden_slot"])
+        self.ave_score = {}
+
+    def convert_intent(self, intent, label_set):
+        """ Convert intent to 0-1 vector.
+        :intent: intent in string format
+        :label_set: an ordered set of intents
+        """
+        vec = np.zeros(len(label_set), dtype=int)
+        for i in intent.split("#"):
+            vec[label_set[i]] = 1
+        return list(vec)
+
+    def convert_slot(self, slot, label_set):
+        """ Convert slot to 0-1 vector.
+        :slot: slot in string format
+        :label_set: an ordered set of slots
+        """
+        vec = np.zeros(len(label_set), dtype=int)
+        vec[label_set[slot]] = 1
+        return list(vec)
+
+    def generate_intent_map(self, intent_list):
+        """ Generate an ordered map for all intents. """
```

Comment:
Can you mention in the comment what the expected form for the intent_list is? (e.g. intents are separated by "#"s)